File: feature-products/src/main/java/com/tirupati/pos/feature/products/data/repository/InsertMockData.kt
// SPDX-FileCopyrightText: 2024-2026 Jayes
// SPDX-License-Identifier: MIT

package com.tirupati.pos.feature.products.data.repository

import com.tirupati.pos.core.database.PendingOperation
import com.tirupati.pos.core.sync.SyncManager
import com.tirupati.pos.core.sync.SyncQueue
import com.tirupati.pos.feature.products.domain.model.Company
import com.tirupati.pos.feature.products.domain.model.Product
import com.tirupati.pos.feature.products.domain.repository.CompanyRepository
import com.tirupati.pos.feature.products.domain.repository.ProductRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.serialization.json.Json
import java.util.UUID
import kotlin.random.Random

/**
 * Utility class to insert mock data into the local database and queue sync operations.
 * This matches the python script mock data schema.
 */
class InsertMockData @javax.inject.Inject constructor(
    private val companyRepository: CompanyRepository,
    private val productRepository: ProductRepository,
    private val syncQueue: SyncQueue,
    private val syncManager: SyncManager
) {
    private val companiesList = listOf(
        "Havells", "Polycab", "Finolex", "Legrand", "Anchor", 
        "Schneider", "Syska", "Philips", "Bajaj", "Crompton", 
        "V-Guard", "Orient", "Usha", "Wipro", "L&T"
    )

    private data class Category(
        val name: String,
        val unit: String,
        val minPrice: Int,
        val maxPrice: Int
    )

    private val categories = listOf(
        Category("Wire 1.0 sq mm", "Coil", 800, 1050),
        Category("Wire 1.5 sq mm", "Coil", 1200, 1500),
        Category("Wire 2.5 sq mm", "Coil", 1800, 2200),
        Category("Wire 4.0 sq mm", "Coil", 2500, 3100),
        Category("Switch 6A", "Pcs", 25, 45),
        Category("Switch 16A", "Pcs", 45, 75),
        Category("Socket 6A", "Pcs", 30, 55),
        Category("Socket 16A", "Pcs", 60, 95),
        Category("MCB 10A SP", "Pcs", 120, 180),
        Category("MCB 16A SP", "Pcs", 120, 180),
        Category("MCB 32A DP", "Pcs", 350, 480),
        Category("MCB 63A DP", "Pcs", 450, 600),
        Category("LED Bulb 9W", "Pcs", 65, 110),
        Category("LED Bulb 12W", "Pcs", 90, 150),
        Category("Ceiling Fan 1200mm", "Pcs", 1400, 1900),
        Category("Exhaust Fan 150mm", "Pcs", 800, 1100),
        Category("PVC Pipe 25mm", "Bundle", 300, 450),
        Category("PVC Pipe 20mm", "Bundle", 250, 380),
        Category("Casing Capping 1 inch", "Bundle", 150, 250),
        Category("Distribution Board 4 Way", "Pcs", 400, 650)
    )

    private val colorsOrTypes = listOf(
        "Red", "Black", "Blue", "Green", "White", 
        "Standard", "Premium", "Heavy Duty", "Gold", "Silver"
    )

    /** Seed the database if it is empty. */
    fun seedIfEmpty() {
        CoroutineScope(Dispatchers.IO).launch {
            // Check if there are any companies already
            val existingCompanies = companyRepository.observeCompanies()
                .firstOrNull() ?: emptyList()
            if (existingCompanies.isNotEmpty()) return@launch

            val seededCompanies = companiesList.map { name ->
                Company(
                    id = UUID.randomUUID().toString(),
                    name = name
                )
            }

            seededCompanies.forEach { company ->
                // Save locally
                companyRepository.saveCompany(company)
                // Queue sync operation
                val op = PendingOperation(
                    id = UUID.randomUUID().toString(),
                    operationType = "INSERT",
                    entityType = "companies",
                    entityId = company.id,
                    payloadJson = Json.encodeToString(Company.serializer(), company),
                    timestamp = System.currentTimeMillis()
                )
                syncQueue.enqueue(op)
            }

            seededCompanies.forEach { company ->
                var productsWritten = 0
                val random = Random(System.currentTimeMillis())
                
                for (cat in categories) {
                    for (variant in colorsOrTypes) {
                        if (productsWritten >= 100) break

                        val pId = UUID.randomUUID().toString()
                        val itemName = "${company.name} ${cat.name} $variant"
                        val itemCode = "${company.name.take(3).uppercase()}-${cat.name.take(3).replace(" ", "").uppercase()}-${variant.take(3).uppercase()}-${String.format("%03d", productsWritten + 1)}"

                        val pRate = random.nextInt(cat.minPrice, cat.maxPrice + 1)
                        val sRate = (pRate * random.nextDouble(1.15, 1.40)).toInt()
                        val stock = random.nextInt(10, 501)

                        val product = Product(
                            id = pId,
                            companyId = company.id,
                            itemCode = itemCode,
                            itemName = itemName,
                            unit = cat.unit,
                            purchaseRate = pRate.toDouble(),
                            sellingRate = sRate.toDouble(),
                            stockQuantity = stock.toDouble()
                        )

                        productRepository.saveProduct(product)

                        val op = PendingOperation(
                            id = UUID.randomUUID().toString(),
                            operationType = "INSERT",
                            entityType = "products",
                            entityId = product.id,
                            payloadJson = Json.encodeToString(Product.serializer(), product),
                            timestamp = System.currentTimeMillis()
                        )
                        syncQueue.enqueue(op)
                        productsWritten++
                    }
                    if (productsWritten >= 100) break
                }
            }
            syncManager.requestSync()
        }
    }
}

File: insert_mock_data.py
import uuid
import time
import random
import requests
import json

SUPABASE_URL = "https://rawcqqejrycunvkqcffu.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InJhd2NxcWVqcnljdW52a3FjZmZ1Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3ODE3NTkzNDAsImV4cCI6MjA5NzMzNTM0MH0.JJdK95dNAhe9UTIp9u7oq4lUClcHcgzqU8cCK7a8d_I"

headers = {
    "apikey": SUPABASE_KEY,
    "Authorization": f"Bearer {SUPABASE_KEY}",
    "Content-Type": "application/json",
    "Prefer": "return=minimal"
}

companies = [
    "Havells", "Polycab", "Finolex", "Legrand", "Anchor", 
    "Schneider", "Syska", "Philips", "Bajaj", "Crompton", 
    "V-Guard", "Orient", "Usha", "Wipro", "L&T"
]

categories = [
    ("Wire 1.0 sq mm", "Coil", 800, 1050),
    ("Wire 1.5 sq mm", "Coil", 1200, 1500),
    ("Wire 2.5 sq mm", "Coil", 1800, 2200),
    ("Wire 4.0 sq mm", "Coil", 2500, 3100),
    ("Switch 6A", "Pcs", 25, 45),
    ("Switch 16A", "Pcs", 45, 75),
    ("Socket 6A", "Pcs", 30, 55),
    ("Socket 16A", "Pcs", 60, 95),
    ("MCB 10A SP", "Pcs", 120, 180),
    ("MCB 16A SP", "Pcs", 120, 180),
    ("MCB 32A DP", "Pcs", 350, 480),
    ("MCB 63A DP", "Pcs", 450, 600),
    ("LED Bulb 9W", "Pcs", 65, 110),
    ("LED Bulb 12W", "Pcs", 90, 150),
    ("Ceiling Fan 1200mm", "Pcs", 1400, 1900),
    ("Exhaust Fan 150mm", "Pcs", 800, 1100),
    ("PVC Pipe 25mm", "Bundle", 300, 450),
    ("PVC Pipe 20mm", "Bundle", 250, 380),
    ("Casing Capping 1 inch", "Bundle", 150, 250),
    ("Distribution Board 4 Way", "Pcs", 400, 650),
]

colors_or_types = ["Red", "Black", "Blue", "Green", "White", "Standard", "Premium", "Heavy Duty", "Gold", "Silver"]

current_time = int(time.time() * 1000)

for c_idx, c_name in enumerate(companies):
    c_id = str(uuid.uuid4())
    
    # Insert company
    c_data = {
        "id": c_id,
        "name": c_name,
        "created_at": current_time,
        "updated_at": current_time
    }
    
    res = requests.post(f"{SUPABASE_URL}/rest/v1/companies", headers=headers, json=[c_data])
    if res.status_code >= 400:
        print(f"Error inserting company {c_name}: {res.text}")
        continue
    
    print(f"Inserted company: {c_name}")
    
    # Generate 100 products
    products_to_insert = []
    products_written = 0
    for cat in categories:
        for variant in colors_or_types:
            if products_written >= 100:
                break
            
            p_id = str(uuid.uuid4())
            cat_name, unit, min_p, max_p = cat
            item_name = f"{c_name} {cat_name} {variant}"
            item_code = f"{c_name[:3].upper()}-{cat_name[:3].upper()}-{variant[:3].upper()}-{products_written+1:03d}"
            
            p_rate = random.randint(min_p, max_p)
            s_rate = int(p_rate * random.uniform(1.15, 1.40))
            stock = random.randint(10, 500)
            
            products_to_insert.append({
                "id": p_id,
                "company_id": c_id,
                "item_code": item_code,
                "item_name": item_name,
                "unit": unit,
                "purchase_rate": p_rate,
                "selling_rate": s_rate,
                "stock_quantity": stock,
                "created_at": current_time,
                "updated_at": current_time
            })
            products_written += 1
        if products_written >= 100:
            break
            
    # Batch insert 100 products
    res = requests.post(f"{SUPABASE_URL}/rest/v1/products", headers=headers, json=products_to_insert)
    if res.status_code >= 400:
        print(f"Error inserting products for {c_name}: {res.text}")
    else:
        print(f"Inserted 100 products for {c_name}")

print("Done inserting mock data into Supabase.")
